### ISSUE TITLE
Feature request GNU like number minus sign in Prolog term parser.

### ISSUE THREAD

**Issue by jburse** [jburse/jekejeke-devel#1]
There is often the question whether Jekejeke Prolog supports the GNU like number minus sign in its Prolog term parser, and if not, when it is planned to support this feature.

The syntax rule under discussion from the ISO core standard is the following rule:

![rule_1](https://cloud.githubusercontent.com/assets/5411487/19974874/4306e930-a1ea-11e6-8442-39477af50aa5.gif)

And the following rule:

![rule_2](https://cloud.githubusercontent.com/assets/5411487/19974880/48852610-a1ea-11e6-86a5-f91bfa2f6ec3.gif)

There are at least two interpretations:
1) Since there is the qualified "directly follows" we have to interpret the integer and float number as directly following without some layout text sequence. Otherwise a clarification would not be necessary, since the production rules say everything.

2) The production rules say everything, its impossible that there are clarfications that modify the producations rules in some way, so as to have an economic presentation of the grammar of the ISO core standard.

These interpretations have both consequences not only for the Prolog term parser, aka read/1 predicate, but also for the Prolog term unparser, aka write/1 predicate. Since we want to be able to safely read back terms that have been written.

In the wild we find both interpretations. The interpretations also carry over how number_codes/2 is implemented. For example SWI-Prolog implements the interpretation 1):

```
Welcome to SWI-Prolog (Multi-threaded, 64 bits, Version 7.3.29)
Copyright (c) 1990-2016 University of Amsterdam, VU Amsterdam

?- X= - 1, write_canonical(X), nl.
-(1)

?- X= -1, write_canonical(X), nl.
-1

?- number_codes(X, "- 1"), write(X), nl.
ERROR: number_codes/2: Syntax error: Illegal number

?- number_codes(X, "-1"), write(X), nl.
-1
```

Whereas GNU-Prolog implements th interpretation 2):

```
GNU Prolog 1.4.4 (64 bits)
Compiled Apr 23 2013, 16:05:07 with cl By Daniel Diaz

?- X = - 1, write_canonical(X), nl.
-1

?- X= -1, write_canonical(X), nl.
-1

?- number_codes(X,"- 1"), write_canonical(X), nl.
-1

?- number_codes(X,"-1"), write_canonical(X), nl.
-1
```

**Comment by jburse** [comment]
In Jekejeke Prolog we follow the interpretation 1), we hence have:

```
Jekejeke Prolog 2, Runtime Library 1.1.6
(c) 1985-2016, XLOG Technologies GmbH, Switzerland

?- X= - 1, write_canonical(X), nl.
-(1)

?- X= -1, write_canonical(X), nl.
-1

?- number_codes(X, "- 1"), write_canonical(X), nl.
Error: Not a number.
- 1
   ^
	number_codes/2

?- number_codes(X, "-1"), write_canonical(X), nl.
-1
```

There no plans whatever to changes this behaviour. What we are planning to do ist open source the parser and unparser of our Prolog system together with the rest of the Prolog system. This might take a couple of months.

When the whole Prolog system is open sourced, one might fork the GitHub repository and change the behaviour of he parser and unparser of our Prolog system. At the moment we have already open sourced the scanner. But the scanner doesn't handle signed numerals:


The handling of interpretation 1) is done in the parser by using the lookAhead() method from the scanner. So its a very efficient and simple method, similar to the way we handle a functor followed by a open parenthesis ("(").  This is only handled in the parser, which is, as aforementioned, not yet open source.

**Comment by UWN** [comment]
Interpretation 1 is invalid. It collides with 6.3.4.2 Operators as functors. To parse
`- 1. ` as a prefix operator, one would have to use one of the two prefix rules for `fx` and `fy. But in **both** the following extra condition is present: 

> Condition: If *a* is a numeric constant, *f* is not `-`

So minus can never be a prefix operator of a numeric constant. You need extra round brackets. In this manner the relatively ambiguous formulation of "followed directly" is indirectly disambiguated.

To summarize: There is no rule to parse `- 1` as a prefix `-` and `1`.

**Comment by jburse** [comment]
The condition doesn't make much sense. Since it does refer to a as an abstract term after parsing. But for example we have also the following rule in the ISO core standard:

![open_close](https://cloud.githubusercontent.com/assets/5411487/19984544/32ad31aa-a20f-11e6-8328-bc89b3d067aa.gif)

So by the same argument of yours the following character sequence - (1) could also not be parsed, since alas the result of parsing (1) is the numerical costant 1.

Possibly the authors of the standard had some good itentions, but things got upside down. Interpretation 1) would imply that the condition on 6.3.4.2 is rephrased as "followed diectly".

**Comment by jburse** [comment]
BTW: GNU Prolog doesn't implement the ISO rule "6.3.4.2 Operators as functors", although in connection with interpretation 2) it would make sense:

Please make this test:
```
GNU Prolog 1.4.4 (64 bits)
Compiled Apr 23 2013, 16:05:07 with cl By Daniel Diaz

?- op(500,fx,-)

?- X = -x*y+z, write_canonical(X), nl.
+(-(*(x,y)),z)

?- X = -2*y+z, write_canonical(X), nl.
+(*(-2,y),z)
```

Expected result according to 6.3.4.2 (Since the abstract term would be *(2,y), right?):
`+(-(*(2,y)),z)
`
Actual result of GNU Prolog:
`+(*(-2,y),z)`

So an interpretation 2) with a correct implementation of ISO grammar rule 6.3.4.2 would make a lot of sense, since it would give what we are used from math notation.

But even GNU Prolog doesn't implement 6.3.4.2 correctly, as the above shows. And of course we should also figure out how to deal with - (1).

In case a method to deal with - (1) is present, a fork of Jekejeke Prolog could implement this. And if everything goes well, maybe it could also become standard for Jekejeke Prolog.

**Comment by UWN** [comment]
Agreed that there is a certain confusion between concrete syntax, abstract syntax, and actual terms. Take `- -1`. Is `-1` here a numeric constant? There is no definition of numeric constant. At least the definition of "integer constant" in 6.1.2 c excludes signs and negative numbers. So is there another abstract entity for such numbers? In 6.3.1.2 Negative numbers the abstract entity *-n* is given, but there is no provision for the meaning of this minus-sign anywhere.

To fix this in the correct way would probably require to split non-terminal `term` into `n term` and `other term` . With `n term = integer | float number`. So negative numbers are part of `other term`. This is a very big change.

However, no matter how the meaning of "numeric constant" gets interpreted, the intention seems to be clear enough: A `-` cannot always serve as prefix, there some conditions where it cannot.  I do not see a way how `- 1.` can now be accepted - when even `- (1).` needs at least some consideration.

Historically, at least since 1978, this particular meaning of `-` has been in DEC10 Prolog (there is some other glitch in the syntax definition of 1978, but as long as operators are not changed it is fine). And from there on in many systems.

Pragmatically, I do not see any system interpreting `- (1).` differently. They all mean `-(1)`. So that is not a big problem. And therefore also the weakness of the current definition is not such a problem.  For `- 1`, the conditions rule out interpreting this as a prefix.

---
(I did not understand the example `-2*y+z`. I do not see how a `(2,y)` can slip in.

**Comment by UWN** [comment]
(Cases like `- 1^2.` are not handled by a non-terminal `n term`.)

Maybe a simpler fix would be to rewrite the Condition for prefix operators:
> Condition: If the first token of *a* is an `integer` or a `float number`.

**Comment by jburse** [comment]
I never said something about `(2,y)`, its `*(2,y)`, please read carefully. Abstract terms are defined in section 6.1.2 of the ISO Prolog standard. According to paragraph e) `*(2,y)` is an abstract term, but they forgot to say that the arguments x_1,..,x_n are abstract terms again. So there are tree omissions in section 6.1.2:
- paragraph c) should also include negated tokens mapped to I.
- paragraph d) should also include negated tokens mapped to F.
- paragraph e) for the construction of the set CT, the arguments must be abstract terms again.

Please note that abstract terms are not syntactic entities. On the syntactic level 0x10 and 16 are different, so are (1) and 1 differnet. On the abstract term level 0x10 and 16 are the same, and (1) and are also the same.  If you change "6.3.4.2 Operators as functors" from abstract terms to syntact level phrases, you change the intent of "6.3.4.2 Operators as functors".

The intent of "6.3.4.2 Operators as functors" is of course to have a small constant evaluation during parsing. So "- 2" is evaluated to -2 as an element of I during parsing. You say the goal of "6.3.4.2 Operators as functors" is to avoid something. I would say its necessary to disambiguate when we do the small constant evaluation and when not, as a partial evaluation technique it shouldn't change the surface syntax at all.

And as a partial evaluation technique, i.e. a small constant evaluation, the following GNU parsing is of course incorrect:
```
?- op(500,fx,-)

?- X = -2*y+z, write_canonical(X), nl.
+(*(-2,y),z)
```
You could fix it, if with your proposal to use a syntactic pharse condition in "6.3.4.2 Operators as functors" instead of an abstract term condition.

On the other hand, for reading mathematical expressions, it could be also appealing to investigate the abstract term condition more closely.

**Comment by UWN** [comment]
You wrote `*(y,2)` which was interpreted as markdown for *(y,2). That's the misunderstanding.

**Comment by jburse** [comment]
One more trap if the meaning of numeric constant in the condition of "6.3.4.2 Operators as functors" is an element of I union F, the parsing "- - 1" would give the abstract term 1.

**Comment by UWN** [comment]
How can this trap happen? 6.3.1.2 Negative numbers are `name, integer ;` only (with name '-'). The next minus can only be a prefix.

**Comment by jburse** [comment]
Well the prefix rule is op, term. And term is "- 1" which gets parsed to the numeric constant -1. Then the prefix rule refuses to do something, because of its condition, and the negative number rule is name, integer which can also not do anything, since in place of integer it has "- 1". Just image a DCG that does backtracking.

**Comment by UWN** [comment]
To my understanding, The entire par in 6.3.1.2 can now be deleted.

**Comment by jburse** [comment]
Yeah you could try to model negative numbers in 6.4, but you will still need a proviso 6.3.4.2. Or do everything in 6.3.4.2. These are two possible alternatives for 6.3.1.2 that come to my mind.

**Comment by UWN** [comment]
Negative numbers are not handled on the token level. Tokenwise, `-1` is two tokens. And the only thing in 6.3.4.2 that needs change is the condition which now would read:

> Condition: If the first token of *a* is an `integer` or a `float number`.

**Comment by jburse** [comment]
We are turning in circles, you already suggested this above. And the "entire par in 6.3.1.2 can now be deleted" was ambiguos to me, I understood you want to delete 6.3.1.2. We are now in SO bla bla mode. Executing my power to delete comments. Watch out!

**Comment by jburse** [comment]
It seems that SWI-Prolog is not the only system with interpretation 1) of the minus sign. Another system that also interprets minus signs this way is ECLiPSe Prolog. This was communicated by Joachim Schimpf on comp.lang.prolog.

Concerning ECLiPSe Prologs discussion of the minus sign, there is also this compilation. The issue of quotes around the minus sign is also discussed. And further we find discussion of the plus sign in relation to numbers:
http://eclipseclp.org/Specs/iso_notes.html#signs

I made some further tests with GNU Prolog:

```
GNU Prolog 1.4.4 (64 bits)
Compiled Apr 23 2013, 17:26:17 with /opt/local/bin/gcc-apple-4.2

?- X = '-' 1, write_canonical(X), nl.
-1

?- X = '-' 1.0e+3, write_canonical(X), nl.
-1000.0

?- number_codes(X, "'-' 1").
uncaught exception: error(syntax_error('constant term stream:1 (char:1) cannot start a number'),number_codes/2)
```

There is some inconsistency, it does not accept the quoted minus sign in number_codes/2.

**Comment by UWN** [comment]
Has there been progress to this?
